Add PX Security operator docs (#976)
portworx/pxdocs: content/shared/cloud-references-security-kubernetes-shared-secret-model-customizing-operator-security.md

---
title: "Customize Security in Portworx"
keywords: authorization, portworx, security, rbac, jwt

> ---

Reviewer: all shared content should have `hidden: true` in the frontmatter

Author: made all of the `cloud-references-security-kubernetes-*` hidden: true.

---
title: "Customize Security in Portworx"
keywords: authorization, portworx, security, rbac, jwt
hidden: true
---

This optional document guides you through customizing your Portworx Operator Security install to fit specific needs.

## Prerequisites

* Portworx Operator 1.4.0 or later
* Portworx Security enabled

### Migrating from Security environment variables to StorageCluster Security spec

In order For Portworx to start with security enabled, it requires a few different environment variables. If you wish to start using the `StorageCluster` security spec, here is how you can migrate your environment variables to spec fields. 

1. First, familiarize yourself with the Security spec in the [StorageCluster](/reference/crd/storage-cluster) article.

2. Create a Kubernetes secret for your shared-secret:
```
kubectl create secret generic px-shared-secret \

>   --from-literal=shared-secret=<Existing Shared Secret Value>

Reviewer 2: Same here. Generate, save in a var, then use

Author: I'm not sure we want to generate this secret in this case. This case is for allowing users to create a k8s secret with their pre-existing shared secret.

If they do not follow this step, we just generate the secret for them.

---

> title: "Customizing Security in Portworx"

Reviewer: ```suggestion
title: "Customize Security in Portworx"
```

keywords: authorization, portworx, security, rbac, jwt
---

This optional document guides you through customizing your Portworx Operator Security install to fit specific needs.

## Prerequisites

* Portworx Operator 1.4.0 or later
* Portworx Security enabled

### Custom issuer, shared secret, and stork key

Add issuer and references to the shared keys as environment variables under `spec.security.auth.selfSigned` of your StorageCluster:

```text
apiVersion: core.libopenstorage.org/v1alpha1
kind: StorageCluster
metadata:
  name: px-cluster
  namespace: kube-system
spec:
  security:
    enabled: true
    auth:
      selfSigned:
        issuer: "portworx.com"
        sharedSecret: "px-shared-secret"
        tokenLifetime: "1h"
  env:
  - name: "PX_SHARED_SECRET"
    valueFrom:
      secretKeyRef:
        name: pxkeys
        key: stork-secret
```

> ### Disabling guest role access

Reviewer: I've been slowly changing these as I go. We should remove gerunds from headings and titles while we're here. It makes them more compact and easier to read when they're in the sidebar or scrollspy. 

```suggestion
### Disable guest role access
```

Starting with Portworx 2.6.0 and later, the [system guest role](/concepts/authorization/overview#guest-access) is enabled by default in order to facilitate easy onboarding to the Portworx Security feature. To turn off this feature, you can simply disable it in the StorageCluster spec:

```text
apiVersion: core.libopenstorage.org/v1alpha1
kind: StorageCluster
metadata:
  name: px-cluster
  namespace: kube-system
spec:
  security:
    enabled: true
    auth:
      guestAccess: 'Disabled'
```

### Managing the guest role yourself

> For fine-grained control over the system.guest role, you can set guestAccess to `Managed`. This will tell the operator to stop updating the [system guest role](/concepts/authorization/overview#guest-access), allowing you to [customize it yourself](/reference/cli/role/#re-enabling-the-system-guest-role).

Reviewer: These tasks should always include a clear instruction for the user. the idea is to separate out the explanation for what this is/does/why you want to do it and the instructions for how exactly to do it:

```suggestion
You can exercise finer control over the system.guest role by setting it `managed` mode. This instructs the Operator to stop updating the [system guest role](/concepts/authorization/overview#guest-access), allowing you to [customize it yourself](/reference/cli/role/#re-enabling-the-system-guest-role).

To enter `managed` mode, set the value of the `spec.security.auth.guestAccess` field to `managed`:
```

review the other sections and see if they can be separated in the same manner.

```text
apiVersion: core.libopenstorage.org/v1alpha1
kind: StorageCluster
metadata:
  name: px-cluster
  namespace: kube-system
spec:
  security:
    enabled: true
    auth:
      guestAccess: 'Managed'
```

### Increasing token lifetime

Additionally, JWT token lifetime can be specified. The operator will generate a token with this token lifetime and refresh it for the user accordingly. 

```text
apiVersion: core.libopenstorage.org/v1alpha1
kind: StorageCluster
metadata:
  name: px-cluster
  namespace: kube-system
spec:
  security:
    enabled: true
    auth:
      tokenLifetime: '4h'
```

### Migrating from environment variables to StorageCluster security spec

> There are a few different environment variables needed to start Portworx with security enabled. If you have previously started Portworx security with environment variables, the same method will work with the operator 1.4 and later. However, if you wish to start using the `StorageCluster` security spec, here are how the environment variables map to spec fields. First, you should familiarize yourself with Security spec in the [StorageCluster](/reference/crd/storage-cluster) article.

Reviewer: Is this saying that you don't need to migrate from env vars to StorageCluster if you did this? I don't think so, so I'm not following this sentence:

> If you have previously started Portworx security with environment variables, the same method will work with the operator 1.4 and later.

I assume this is what the steps show you how to do? If so, we should be more explicit about that. 
> However, if you wish to start using the `StorageCluster` security spec, here are how the environment variables map to spec fields.

Why should they familiarize themselves with this spec? what's the relation between that spec and here?

>First, familiarize yourself with the Security spec in the [StorageCluster](/reference/crd/storage-cluster) article.

```suggestion
In order For Portworx to start with security enabled, it requires a few different environment variables. If you have previously started Portworx security with environment variables, the same method will work with the operator 1.4 and later. However, if you wish to start using the `StorageCluster` security spec, you must map the environment variables to spec fields. 

First, familiarize yourself with the Security spec in the [StorageCluster](/reference/crd/storage-cluster) article.
```

---
title: "Customize Security in Portworx"
keywords: authorization, portworx, security, rbac, jwt
hidden: true
---

> This optional document guides you through customizing your Portworx Operator Security install to fit specific needs.

Author: Having optional in the title felt crowded in the nav bar. I'll move it to be after the example app.

## Prerequisites

> * Portworx Operator 1.4.0 or later

Reviewer 3: Just use 1.4 to be consistent with other places we have mention 1.4 operator.

* Portworx Security enabled

### Add a custom issuer, shared secret, and stork key to your StorageCluster

Add issuer and references to the shared keys as environment variables under `spec.security.auth.selfSigned` of your StorageCluster:

```text
apiVersion: core.libopenstorage.org/v1alpha1
kind: StorageCluster
metadata:
  name: px-cluster
  namespace: kube-system
spec:
  security:
    enabled: true
    auth:
      selfSigned:
        issuer: "portworx.com"
        sharedSecret: "px-shared-secret"
        tokenLifetime: "1h"
  ...
  stork:
    env:

>     - name: "PX_SHARED_SECRET"

Author: Yes, good point
